### PR TITLE
Add create_extensible_enum macro

### PR DESCRIPTION
**Pull request by heaths**

### CODE REVIEW

**Comment by heaths** [comment]
This uses a little "trick" of characters really only usable in macros to effectively make a match arm "internal". We can use this for shared code for DRY. I probably could've gone a little further with some creative match arms, but I think that'd diminish readability and these macros are hopefully temporary anyway (see #1846).

File: sdk/typespec/typespec_client_core/src/macros.rs [hunk 1]
```diff
@@ -189,27 +290,63 @@ mod test {
     }
 
     #[test]
-    fn test_color_parse_1() {
+    fn color_parse_1() {
```
Comment (heaths) [inline sdk/typespec/typespec_client_core/src/macros.rs:293]: Note: I don't like "test_" prefixes when in a "test(s)" module. It's redundant. "test_" is great when tests are in the same module e.g.,

```rust
fn foo() {}

#[test]
fn test_foo() {}
```

**Review by jhendrixMSFT — APPROVED** [review]

**Review by analogrelay — APPROVED** [review]

**Review by LarryOsterman — APPROVED** [review]
I'm annoyed by the redundancy associated with having to say `Black, "Black"` and the potential for subtle bugs, but absent adding the ability for rust to introduce a tokenizing operator, I don't see a way of avoiding it.

**Comment by heaths** [comment]
> I'm annoyed by the redundancy associated with having to say `Black, "Black"` and the potential for subtle bugs, but absent adding the ability for rust to introduce a tokenizing operator, I don't see a way of avoiding it.

It actually wouldn't be that hard and I had considered it, but it's of dubious value because 1) these are temporary, and 2) they wouldn't likely be used in emitters because it'd be harder - and rather pointless - to vary the emitted code. I suppose, also, 3) it's not that much more to type in what few hand-coded cases there should be but, again, see point 1.

**Pull request merged**